File: ios-native/KPKNFit/KPKNFit/Domain/Auge/DiscomfortAggregationEngine.swift
import Foundation

struct SessionDiscomfortSummary {
    let discomfortId: String
    let label: String
    let reportedInExercises: [String]
    let articularVolumeScore: Int
    let relatedArticular: [ArticularBattery]
}

enum DiscomfortAggregationEngine {

    static func computeSessionDiscomfortSummary(
        postExerciseFeedbackByExerciseId: [String: PostExerciseFeedback],
        completedExercises: [CompletedExercise],
        exerciseDb: [String: ExerciseMuscleInfo]
    ) -> [SessionDiscomfortSummary] {
        var articularSetCounts: [ArticularBattery: Int] = [:]

        for exercise in completedExercises {
            let workingSets = exercise.sets.filter { !$0.isWarmup && !$0.skipped }.count
            if workingSets == 0 { continue }

            let dbInfo = (exercise.exerciseDbId ?? exercise.exerciseId)
                .flatMap { exerciseDb[$0.lowercased()] }

            guard let dbInfo = dbInfo else { continue }

            let relatedArticulars = Array(Set(
                dbInfo.involvedMuscles.flatMap { AugeTtcEngine.MUSCLE_TO_ARTICULAR[$0.muscle] ?? [] }
            ))

            for articular in relatedArticulars {
                articularSetCounts[articular, default: 0] += workingSets
            }
        }

        let flatDiscomforts = postExerciseFeedbackByExerciseId
            .filter { _, feedback in feedback.discomfortIds.contains { $0 != "none" } }
            .flatMap { exerciseId, feedback in
                feedback.discomfortIds
                    .filter { $0 != "none" }
                    .map { (discomfortId: $0, exerciseId: exerciseId) }
            }

        let discomfortExerciseMap = Dictionary(grouping: flatDiscomforts, by: { $0.discomfortId })
            .mapValues { items in items.map { $0.exerciseId } }

        var summaries: [SessionDiscomfortSummary] = []
        for (discomfortId, exerciseIds) in discomfortExerciseMap {
            let entry = DISCOMFORT_CATALOG_BY_ID[discomfortId]
            let label = entry?.label ?? discomfortId
            let relatedArticular = entry?.relatedArticular ?? []

            let volumeScore = relatedArticular.reduce(0) { acc, articular in
                acc + (articularSetCounts[articular] ?? 0)
            }

            let exerciseNames = exerciseIds.compactMap { id in
                completedExercises.first { $0.exerciseId == id }?.exerciseName
            }.uniqued()

            summaries.append(SessionDiscomfortSummary(
                discomfortId: discomfortId,
                label: label,
                reportedInExercises: exerciseNames,
                articularVolumeScore: volumeScore,
                relatedArticular: relatedArticular
            ))
        }

        return summaries
            .sorted { $0.articularVolumeScore > $1.articularVolumeScore }
            .prefix(5)
            .map { $0 }
    }
}

private extension Array where Element: Hashable {
    func uniqued() -> [Element] {
        var seen = Set<Element>()
        return filter { seen.insert($0).inserted }
    }
}

private extension Dictionary {
    func grouped<Key: Hashable>(by keyTransform: (Value) -> Key, valueTransform: (Value) -> Value) -> [Key: [Value]] {
        var result: [Key: [Value]] = [:]
        for value in values {
            let key = keyTransform(value)
            result[key, default: []].append(valueTransform(value))
        }
        return result
    }
}

File: android-native/app/src/main/java/com/example/kpkn/domain/auge/DiscomfortAggregationEngine.kt
package com.example.kpkn.domain.auge

import com.example.kpkn.data.models.ArticularBattery
import com.example.kpkn.data.models.CompletedExercise
import com.example.kpkn.data.models.DISCOMFORT_CATALOG_BY_ID
import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.PostExerciseFeedback

data class SessionDiscomfortSummary(
    val discomfortId: String,
    val label: String,
    val reportedInExercises: List<String>,
    val articularVolumeScore: Int,
    val relatedArticular: List<ArticularBattery>,
)

object DiscomfortAggregationEngine {

    fun computeSessionDiscomfortSummary(
        postExerciseFeedbackByExerciseId: Map<String, PostExerciseFeedback>,
        completedExercises: List<CompletedExercise>,
        exerciseDb: Map<String, ExerciseMuscleInfo>,
    ): List<SessionDiscomfortSummary> {
        val articularSetCounts = mutableMapOf<ArticularBattery, Int>()

        for (exercise in completedExercises) {
            val workingSets = exercise.sets.count { !it.isWarmup && !it.skipped }
            if (workingSets == 0) continue

            val dbId = exercise.catalogConfigurationId ?: exercise.exerciseDbId ?: exercise.exerciseId
            val dbInfo = dbId?.trim()?.lowercase()?.let(exerciseDb::get)
            if (dbInfo == null) continue

            val involvedMuscles = exercise.effectiveMuscles?.takeIf { it.isNotEmpty() }
                ?: dbInfo.involvedMuscles
            val relatedArticulars = involvedMuscles
                .flatMap { im -> AugeTtcEngine.articularBatteriesFor(im.muscle, im.emphasis) }
                .distinct()

            for (articular in relatedArticulars) {
                articularSetCounts[articular] = (articularSetCounts[articular] ?: 0) + workingSets
            }
        }

        val discomfortExerciseMap = postExerciseFeedbackByExerciseId
            .filter { (_, feedback) -> feedback.discomfortIds.any { it != "none" } }
            .flatMap { (exerciseId, feedback) ->
                feedback.discomfortIds
                    .filter { it != "none" }
                    .map { it to exerciseId }
            }
            .groupBy({ it.first }, { it.second })

        val summaries = discomfortExerciseMap.map { (discomfortId, exerciseIds) ->
            val entry = DISCOMFORT_CATALOG_BY_ID[discomfortId]
            val label = entry?.label ?: discomfortId
            val relatedArticular = entry?.relatedArticular.orEmpty()

            val volumeScore = relatedArticular.sumOf { articular ->
                articularSetCounts[articular] ?: 0
            }

            val exerciseNames = exerciseIds.mapNotNull { id ->
                completedExercises.firstOrNull { it.exerciseId == id }?.exerciseName
            }.distinct()

            SessionDiscomfortSummary(
                discomfortId = discomfortId,
                label = label,
                reportedInExercises = exerciseNames,
                articularVolumeScore = volumeScore,
                relatedArticular = relatedArticular,
            )
        }

        return summaries
            .sortedByDescending { it.articularVolumeScore }
            .take(5)
    }
}
